crates/pera-core/src/signature_mpc/mpc_manager.rs

```diff
@@ -0,0 +1,450 @@
+use crate::authority::authority_per_epoch_store::AuthorityPerEpochStore;
+use crate::consensus_adapter::{ConsensusAdapter, SubmitToConsensus};
+use crate::signature_mpc::mpc_events::{CreatedProofMPCEvent, MPCEvent};
+use crate::signature_mpc::proof::ProofParty;
+use anyhow::anyhow;
+use group::secp256k1::group_element::Value;
+use group::{secp256k1, GroupElement, PartyID};
+use im::hashmap;
+use itertools::Itertools;
+use mpc::party::Advance;
+use mpc::{two_party::Round, AdvanceResult};
+use pera_types::base_types::{AuthorityName, ObjectID, PeraAddress};
+use pera_types::error::{PeraError, PeraResult};
+use pera_types::event::Event;
+use pera_types::messages_consensus::ConsensusTransaction;
+
+use pera_types::committee::EpochId;
+use rand_core::OsRng;
+use rayon::prelude::*;
+use schemars::_private::NoSerialize;
+use serde::{Deserialize, Serialize};
+use std::cmp::PartialEq;
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::future::Future;
+use std::io::Write;
+use std::marker::PhantomData;
+use std::sync::{Arc, Weak};
+use std::time::Duration;
+use std::{io, mem};
+use tokio::sync::{mpsc, Mutex, RwLock};
+use tokio::time::sleep;
+use tracing::{debug, error, info};
+
+/// The message a validator can send to the other parties while running a signature MPC session.
+#[derive(Clone)]
+struct SignatureMPCMessage {
+    /// The serialized message
+    message: Vec<u8>,
+    /// The authority that sent the message
+    authority: AuthorityName,
+}
+
+/// A wrapper for the generic Party trait that allows creating new instances of the Party from only the threshold.
+/// Should be implemented internally in newer versions of the [`proof`] crate.
+pub trait CreatableParty: Advance {
+    /// The MPC Manager will create a new mpc instance after the init event is received.
+    type InitEvent: MPCEvent + Serialize + for<'a> Deserialize<'a>;
+    /// The MPC Manager will finalize the mpc instance after the finalize event is received.
+    type FinalizeEvent: MPCEvent + Serialize + for<'a> Deserialize<'a>;
+
+    fn new(parties: HashSet<PartyID>, party_id: PartyID) -> Self;
+}
```

Comment:
```suggestion
```
With the latest code, we don't need this anymore.
Instead, I have either `Defualt` implemented, or a `From<DecryptionKeyShare>`.
Other complexity is moved to the `AuxiliaryInput`. See `2pc_mpc::Protocol`

Reply:
As we agreed we'll change this in a later PR

Reply:
Done